### PR TITLE
Configurable & optimized memory management

### PR DESCRIPTION
**Pull request by antonfirsov**
### Prerequisites
- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/SixLabors/ImageSharp/pulls) open
- [x] I have verified that I am following matches the existing coding patterns and practice as demonstrated in the repository. These follow strict Stylecop rules :cop:.
- [x] I have provided test coverage for my change (where applicable)

## PR summary
- Finished the changes started by @rytmis on #431
- Enables configurable memory management as described in #225
- With the default pooling settings, ImageSharp memory usage is approximately halved.

## Changes
- Implemented all the *beta-3* points of #225 
  - There is no `MemoryManager.Return()` in the final design, because the buffer implementations are now [bound]([replaced](https://github.com/SixLabors/ImageSharp/blob/feature/memory-manager/src/ImageSharp/Memory/ArrayPoolMemoryManager.Buffer%7BT%7D.cs#L46)) to their `MemoryManager`-s.
- A workaround for [coreclr/#16443](https://github.com/dotnet/coreclr/issues/16443), hope it will fix travis builds
- [A few regression tests](https://github.com/SixLabors/ImageSharp/blob/feature/memory-manager/tests/ImageSharp.Tests/Drawing/SolidBezierTests.cs#L37) for drawing methods. (Needed it to allow changes to be made safely.)

## Benchmarks
[Load testing benchmarks](https://github.com/SixLabors/ImageSharp/tree/load-test/tests/ImageSharp.LoadTest#benchmark-scenario) have been implemented on the `load-test` branch. (The  project should be moved into it's own SixLabors repository later). [Link to results](https://1drv.ms/x/s!AkF1IWe7aWXrm17hrGM1htnxT1if)

#### TLDR
This is how (managed/GC) memory consumption changes for 10,000 resize requests of ~3200p Jpegs, new defaults VS pre-MemoryManager (old):
image
The peaks are measured around requests with big outlier images.

#### Throughput
The throughput visualized for different scenarios & pool settings (measured in MegaPixel/sec):
image

### Memory footprint
Avarage managed memory footprint while serving ~600 resize requests (different scenarios & pool settings):
image

#### Conclusions
- With the default pooling settings, ImageSharp memory usage is approximately halved. I hope these settings are now good enough for most users.
- If your environment is memory-constrained:
  - Use `ArrayPoolMemoryManager.CreateWithModeratePooling()` or `.CreateWithMinimalPooling()`! For these two: the difference in memory consumption is not as big as the difference in throughput.
  - Do not use `SimpleGCMemoryManager` ("NullMemoryManager"), it only makes things worse!
- For servers under high load `ArrayPoolMemoryManager.CreateWithAggressivePooling()` can increase the throughput with the cost of a higher memory footprint
- The memory usage is always being stabilized after a warm-up period.
- The throughput of the pre-MemoryManager ("Old") version is significantly higher. Not sure if the performance regression is caused directly by changes in memory management. We need to investigate this.

/cc @vpenades @GeorgePlotnikov @denisivan0v

### CODE REVIEW

**Comment by antonfirsov** [comment]
What's wrong with the CI builds? 😥 I haven't even touched those lines in `AffineTransformProcessor.cs`.

**Comment by vpenades** [comment]
@antonfirsov Thanks for the great effort you've put on this!

So, from the code I've had the time to review, I undestand that any resource created without an explicit configuration will use the default MemoryManager set in the default, static, Configuration.

So, to use an alternative memory manager, there's two choices:
- Explicitly set a memory manager to each created resource.
- Set Configuration.Default.MemoryManager prior to any other operation

Is that right?

About your memory stress tests and the suggestion of never using SimpleGCMemoryManager; I can see that suggestion comes from the fact that SimpleGCMemoryManager performs slower, and consumes more memory than any of the other pooled memory managers...  under the assumption that you run ImageSharp non stop (web service scenario).

But that was never the point of that null memory manager; I would like to see what happens with all the memory managers with this alternative scenario (client application scenario)

1. Run some resizes, like with the current stress tests.
2. Stop using ImageSharp, and run a full GC collection, one that also tries to free any large object.
3. Test result: how much memory was actually freed, and how much is still locked by the array pools?

This scenario is more consistent with applications doing image processing sparsely, and then doing something completely different that would require as much memory available as possible.

**Comment by antonfirsov** [comment]
>    Run some resizes, like with the current stress tests.
>    Stop using ImageSharp, and run a full GC collection, one that also tries to free any large object.
>    Test result: how much memory was actually freed, and how much is still locked by the array pools?

@vpenades as you wish!
On this chart you can see the effect of calling `MemoryManager.ReleaseRetainedResources()` + `GC.Collect()` after ~200 requests:
image

> So, to use an alternative memory manager, there's two choices: [...]

Yes you always need to explicitly pass a memory manager to individual operations. I suggest yo to implement an `ImagingContext` class which owns a  `MemoryManager` for you and contains your commonly used factory methods & operations. As @JimBobSquarePants explained, we did not want such a thing to be part of our core API, because it complicates things, and hardens the learning curve of the library, while not needed for the 90% of our users.

**Comment by vpenades** [comment]
@antonfirsov Upon a second review, I've realized the Configuration class essentially handles memory managers and image formats, but Image Formats are handled directly by the configuration class.

Maybe it could be cleaner to have a ImageFormatsCollection that keeps all the functionality related to image formats management, and Configuration greatly simplified. That would make life easier when creating multiple configurations while keeping a shared collection of image formats.

Something like this:

```
class Configuration
{
    public ParallelOptions ParallelOptions {get;set;
    public MemoryManager Manager {get; set;}
    public FormatsCollection Formats {get; set;}
}
```

File: src/ImageSharp/Memory/BufferArea{T}.cs [hunk 1]
```diff
@@ -45,11 +45,26 @@ public BufferArea(IBuffer2D<T> destinationBuffer)
...
         public int Stride => this.DestinationBuffer.Width;
 
+        /// <summary>
+        /// Gets a value indicating whether the area refers to the entire <see cref="DestinationBuffer"/>
+        /// </summary>
+        public bool IsFullBufferArea => this.Size == this.DestinationBuffer.Size();
+
         /// <summary>
         /// Gets or sets a value at the given index.
         /// </summary>
```
Comment (JimBobSquarePants) [inline src/ImageSharp/Memory/BufferArea{T}.cs]: *Diff wouldn't let me highlight where I wanted to*

`GetReferenceToOrigo` I'm assuming this is a typo? `GetReferenceToOrigin` perhaps?

File: src/ImageSharp/Formats/Bmp/BmpDecoderCore.cs [hunk 2]
```diff
@@ -221,9 +224,9 @@ private void ReadRle8<TPixel>(PixelAccessor<TPixel> pixels, byte[] colors, int w
             var color = default(TPixel);
             var rgba = new Rgba32(0, 0, 0, 255);
 
-            using (var buffer = Buffer2D<byte>.CreateClean(width, height))
+            using (var buffer = this.configuration.MemoryManager.Allocate2D<byte>(width, height, true))
```
Comment (JimBobSquarePants) [inline src/ImageSharp/Formats/Bmp/BmpDecoderCore.cs]: `AllocateClean2D`?

**Comment by JimBobSquarePants** [comment]
>Maybe it could be cleaner to have a image formats decoupled from Configurationinto it's own class ImageFormatsCollection that would keeps all the functionality related to image formats management, and would let Configuration to be greatly simplified.

@vpenades Looking at `Configuration` class I'm inclined to agree. A separate PR though I think. 

@antonfirsov Re throughput performance? Perhaps it's as simple as the abstraction costs due to using interfaces? I can't see anything that stands out.

File: src/ImageSharp/Formats/Jpeg/PdfJsPort/Components/PdfJsHuffmanTable.cs [hunk 3]
```diff
@@ -12,32 +12,34 @@ namespace SixLabors.ImageSharp.Formats.Jpeg.PdfJsPort.Components
...
         /// <param name="lengths">The code lengths</param>
         /// <param name="values">The huffman values</param>
-        public PdfJsHuffmanTable(byte[] lengths, byte[] values)
+        public PdfJsHuffmanTable(MemoryManager memoryManager, byte[] lengths, byte[] values)
         {
-            this.lookahead = Buffer<short>.CreateClean(256);
-            this.valOffset = Buffer<short>.CreateClean(18);
-            this.maxcode = Buffer<long>.CreateClean(18);
+            // TODO: Replace FakeBuffer<T> usages with standard or array orfixed-sized arrays
+            this.lookahead = memoryManager.AllocateFake<short>(256);
```
[the remark this answers is not in the document]
Comment (antonfirsov) [reply src/ImageSharp/Formats/Jpeg/PdfJsPort/Components/PdfJsHuffmanTable.cs]: Yeah, that would allow us to get rid of the `AllocateFake<T>` workaround entirely.

File: src/ImageSharp/Advanced/AdvancedImageExtensions.cs [hunk 4]
```diff
@@ -88,6 +88,14 @@ internal static Span<TPixel> GetPixelRowSpan<TPixel>(this Image<TPixel> source,
             where TPixel : struct, IPixel<TPixel>
             => source.Frames.RootFrame.GetPixelRowSpan(row);
 
+        /// <summary>
```
[the remark this answers is not in the document]
Comment (antonfirsov) [reply src/ImageSharp/Advanced/AdvancedImageExtensions.cs]: We need `Memory<T>` for the memory interop stuff, for the rest we need a proper API review. Not sure if stuff like `.GetMemoryManager()` should be public.

File: src/ImageSharp/Memory/BufferArea{T}.cs (hunk 1 above)
Comment (antonfirsov) [reply src/ImageSharp/Memory/BufferArea{T}.cs]: It's just my Hunglish :D

File: src/ImageSharp.Drawing/Paths/ShapeRegion.cs [hunk 5]
```diff
@@ -46,18 +46,17 @@ public override int Scan(float y, float[] buffer, int offset)
...
-            {
-                PointF[] array = innerBuffer.Array;
-                int count = this.Shape.FindIntersections(start, end, array, 0);
 
-                for (int i = 0; i < count; i++)
-                {
-                    buffer[i + offset] = array[i].X;
-                }
+            // TODO: This is a temporal workaround because of the lack of Span<T> API-s on IPath. We should use MemoryManager.Allocate() here!
+            PointF[] innerBuffer = new PointF[buffer.Length];
```
[the remark this answers is not in the document]
Comment (antonfirsov) [reply src/ImageSharp.Drawing/Paths/ShapeRegion.cs]: Should we do it now, or in a follow up PR pair?

**Comment by antonfirsov** [comment]
@JimBobSquarePants a quick initial investigation regarding the regression in speed:
Looks like there are several factors behind that, scattered through the whole codebase.
The most significant is probably `Span<T>` indexer, and the use of `IBuffer<T>.Span` (accessing elements in a lazy way: `buffer.Span[i]`). Span is being used now on many places where we've been using locally pooled arrays earlier, it has a high impact therefore.

I've been running my benchmarks on .NET 4.6.1 runtime (was simpler to work with command line), where indexing `Span<T>` is significantly slower than indexing arrays. The slowdown of the `ReadByteStuffedByteUnsafe()` is a quite trivial example in the Jpeg decoder:
image

Like it or not, but if we want the library to be fast on the classic .NET, we have to replace indexers with `Unsafe.Add(ref base, i)` on hot path.

I will continue my investigation and maybe apply fixes on the weakest points.

File: src/ImageSharp.Drawing/Paths/ShapeRegion.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (JimBobSquarePants) [reply src/ImageSharp.Drawing/Paths/ShapeRegion.cs]: Let's do a follow up once we've done Shapes.

File: src/ImageSharp/Formats/Jpeg/PdfJsPort/Components/PdfJsHuffmanTable.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (JimBobSquarePants) [reply src/ImageSharp/Formats/Jpeg/PdfJsPort/Components/PdfJsHuffmanTable.cs]: 👍

File: src/ImageSharp/Advanced/AdvancedImageExtensions.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (JimBobSquarePants) [reply src/ImageSharp/Advanced/AdvancedImageExtensions.cs]: Cool, I'm looking forward to MS getting that released.

File: src/ImageSharp/Memory/BufferArea{T}.cs (hunk 1 above)
Comment (JimBobSquarePants) [reply src/ImageSharp/Memory/BufferArea{T}.cs]: 😄 Haha... Far better than any language skills I've accumulated.

**Comment by JimBobSquarePants** [comment]
@antonfirsov I'm happy for this to be merged now. 👍 

`Unsafe<T>` is fantastically readable really (It's such a great API!) so I have no issue replacing indexer access on hot paths with that. That can be done as small individual PR's though as we benchmark after the beta. (I'm really hoping the final push will be all about optimization with very few bug fixes/changes)

**Comment by antonfirsov** [comment]
@JimBobSquarePants check out my latest optimization commits. Pushed stuff that would normally deserve a separate explanatory PR.

Maybe I'll add a few more of these (only for stuff that's covered!), and we can merge afterwards.

**Comment by JimBobSquarePants** [comment]
`WithTemporalBuffer<T>` is very nice! I really should get better at reading up on the API's I use, I wasn't aware of the before/after action overloads to `Parallel.For`

Ok, rock on, I'll leave merging to you.

**Comment by antonfirsov** [comment]
Okkay, I guess I managed to bring back speed to previous levels with just a few changes, so no one would feel there's a regression.

**Comment by JimBobSquarePants** [comment]
That's a neat trick you used to access the fixed `Data` field in the unfixed indexer expression. I'll have to remember that one.

File: src/ImageSharp/Formats/Jpeg/GolangPort/Components/Decoder/Bytes.cs [hunk 6]
```diff
@@ -58,10 +58,12 @@ internal struct Bytes : IDisposable
         /// <returns>The bytes created</returns>
         public static Bytes Create(MemoryManager memoryManager)
```
Comment (JimBobSquarePants) [inline src/ImageSharp/Formats/Jpeg/GolangPort/Components/Decoder/Bytes.cs]: No need to pass the parameter here now or to the InputProcessor.

Comment (antonfirsov) [reply src/ImageSharp/Formats/Jpeg/GolangPort/Components/Decoder/Bytes.cs]: Thanks for fixing it!

**Comment by antonfirsov** [comment]
@JimBobSquarePants 
Re `Parallel.For` parameters:
I also discovered them yesterday! I was googling for that feature (I was quite sure it *must* exist), and it was right here on the `Parallel.For`, API we just failed to notice it :)

Re indexable fixed structs:
It's quite an unsafe trick, so one should be very careful using it. Strictly speaking, it should be named something like `GetItemUnsage(idx)` instead of simply being an indexer, but that naming is soooo verbose.

**Comment by vpenades** [comment]
> @vpenades Looking at Configuration class I'm inclined to agree. A separate PR though I think.

@JimBobSquarePants Yes, please, this PR is a bit useless if creating manual configurations is non trivial.

**Comment by antonfirsov** [comment]
I'm merging this now, further optimizations and API changes will be added as separate PR-s.
